cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandlerInternal.java

```diff
@@ -200,4 +244,125 @@ public void getAppDetailForVersion(HttpRequest request, HttpResponder responder,
+          try {
+            ApplicationWithPrograms app = applicationLifecycleService.deployApp(appId, appRequest,
+                null, createProgramTerminator(), skipMakingLatest);
+            responder.sendJson(HttpResponseStatus.OK, GSON.toJson(getApplicationRecord(app)));
+          } catch (DatasetManagementException e) {
+            if (e.getCause() instanceof UnauthorizedException) {
+              throw (UnauthorizedException) e.getCause();
+            } else {
+              throw e;
+            }
+          }
+        } catch (ArtifactNotFoundException e) {
+          responder.sendString(HttpResponseStatus.NOT_FOUND, e.getMessage());
+        } catch (ConflictException e) {
+          responder.sendString(HttpResponseStatus.CONFLICT, e.getMessage());
+        } catch (UnauthorizedException e) {
+          responder.sendString(HttpResponseStatus.FORBIDDEN, e.getMessage());
+        } catch (InvalidArtifactException e) {
+          responder.sendString(HttpResponseStatus.BAD_REQUEST, e.getMessage());
+        } catch (IOException e) {
+          LOG.error("Error reading request body for creating app {}.", appId, e);
+          responder.sendString(HttpResponseStatus.INTERNAL_SERVER_ERROR, String.format(
+              "Error while reading json request body for app %s.", appId));
+        } catch (Exception e) {
+          LOG.error("Deploy failure", e);
+          responder.sendString(HttpResponseStatus.BAD_REQUEST, e.getMessage());
+        }
+      }
+    };
+  }
+
+  /**
+   * Mark provided application versions as latest
+   *
+   * @param request {@link FullHttpRequest}
+   * @param responder {@link HttpResponse}
+   * @param namespace the namespace where the applications are deployed
+   * @throws Exception if either namespace or any of the application (versions) doesn't exist
+   */
+  @POST
+  @Path("/apps/markLatest")
+  public void markApplicationsAsLatest(FullHttpRequest request, HttpResponder responder,
+      @PathParam("namespace-id") final String namespace) throws Exception {
+    NamespaceId namespaceId = new NamespaceId(namespace);
+    if (!namespaceQueryAdmin.exists(namespaceId)) {
+      throw new NamespaceNotFoundException(namespaceId);
+    }
+
+    MarkLatestAppsRequest appsMarkLatestRequest;
+    try {
+      appsMarkLatestRequest = parseBody(request, MarkLatestAppsRequest.class);
+    } catch (JsonSyntaxException e) {
+      throw new BadRequestException("Invalid request body: " + e.getMessage());
+    }
+
+    if (appsMarkLatestRequest == null) {
+      throw new BadRequestException("Invalid request body.");
+    }
+
+    List<ApplicationId> appIds = new ArrayList<>();
+    for (AppVersion appRequest : appsMarkLatestRequest.getApps()) {
+      appIds.add(new ApplicationId(namespace, appRequest.getName(), appRequest.getAppVersion()));
+    }
+
+    try {
+      applicationLifecycleService.markAppsAsLatest(appIds);
+      responder.sendString(HttpResponseStatus.OK, "Success");
+    } catch (IOException e) {
```

Comment:
Have commented multiple times on this, these exception should not be caught, they should propagate up and be handled by the common exception handler code.

Reply:
done.